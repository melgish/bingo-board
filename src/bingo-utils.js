import gen from "random-seed"
import shuffle from 'shuffle-array';

export const GAME = 'COVID';

// Array of indexes 0 - 4 because I'm that lazy
const BINGO = GAME.split('').map((v, k) => k);
const FREE = 2;
const RNG = gen.create();

// Array of all available bingo numbers arranged in board layout
// [1, 2, ... 15]
// [16, 17, ... 30]
// ...
export const BOARD = BINGO.map(
  i => Array.from({length: 15}, (v, k) => (i * 15) + k + 1)
);


/**
 * Create new 5 digit seed
 *
 * @return {number} random 5 digit number
 */
export function getSeed() {
  return RNG.intBetween(10000, 100000);
}

/**
 * Generate and return card for given seed
 *
 * @param {number} seed
 */
export function getCard(seed) {
  const picks = 5;
  // Use repeatable random number generator.
  const rng = gen.create(seed).random;
  // Capture 5 random values from each letter
  const cols = BINGO.map(i => shuffle.pick(BOARD[i], { picks, rng }));
  // Clear the free space
  cols[FREE][FREE] = 0;
  // Arrange in rows
  const rows = BINGO.map(i => BINGO.map(j => cols[j][i]));

  return { seed, rows };
}

