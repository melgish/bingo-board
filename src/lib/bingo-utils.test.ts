import { describe, expect, it } from "vitest"
import { getCard, getSeed } from "./bingo-utils"

describe("getSeed", () => {
  it("should return a random number", () => {
    const seed = getSeed()
    expect(seed).toBeGreaterThan(9999)
    expect(seed).toBeLessThan(100000)
  })
})

describe("getCard", () => {
  it("should create a bingo card", () => {
    expect(getCard(12345)).toMatchSnapshot()
  })
})
