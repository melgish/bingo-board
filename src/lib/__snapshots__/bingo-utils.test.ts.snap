// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getCard > should create a bingo card 1`] = `
{
  "rows": [
    [
      12,
      18,
      34,
      47,
      66,
    ],
    [
      13,
      16,
      32,
      53,
      62,
    ],
    [
      4,
      29,
      0,
      50,
      70,
    ],
    [
      7,
      23,
      37,
      56,
      72,
    ],
    [
      2,
      30,
      31,
      49,
      63,
    ],
  ],
  "seed": 12345,
}
`;
